src/test/java/com/google/api/pathtemplate/PathTemplateTest.java

```diff
@@ -171,6 +172,250 @@ public void matchWithUnboundInMiddle() {
+    PathTemplate template = PathTemplate.create("projects/{project}/zones/{zone_a}~{zone_b}");
+    Truth.assertThat(template.match("projects/project-123/zones/europe-west3-c")).isNull();
+
+    template = PathTemplate.create("projects/{project}/zones/{zone_a}~{zone_b}.{zone_c}");
+    Map<String, String> match =
+        template.match("projects/project-123/zones/europe-west3-c~.us-east3-a");
+    Truth.assertThat(match).isNotNull();
+    Truth.assertThat(match.get("project")).isEqualTo("project-123");
+    Truth.assertThat(match.get("zone_a}~{zone_b")).isNull();
+    Truth.assertThat(match.get("zone_a")).isEqualTo("europe-west3-c");
+    Truth.assertThat(match.get("zone_b")).isEmpty();
+    Truth.assertThat(match.get("zone_c")).isEqualTo("us-east3-a");
+  }
+
+  @Test
+  public void complexResourceIdNoSeparator() {
+    thrown.expect(ValidationException.class);
+    PathTemplate.create("projects/{project}/zones/{zone_a}{zone_b}");
+    thrown.expectMessage(
+        String.format(
+            "parse error: missing or invalid complex resource ID delimiter character in '%s'",
+            "{zone_a}{zone_b}"));
+
+    PathTemplate.create("projects/{project}/zones/{zone_a}_{zone_b}{zone_c}");
+    thrown.expectMessage(
+        String.format(
+            "parse error: missing or invalid complex resource ID delimiter character in '%s'",
+            "{zone_a}_{zone_b}{zone_c}"));
+  }
+
+  @Test
+  public void complexResourceIdInvalidDelimiter() {
+    thrown.expect(ValidationException.class);
+    // Not a comprehensive set of invalid delimiters, please check the class's defined pattern.
+    List<String> someInvalidDelimiters =
+        new ArrayList<>(Arrays.asList("|", "!", "@", "a", "1", ",", "{", ")"));
+    for (String invalidDelimiter : someInvalidDelimiters) {
+      PathTemplate.create(
+          String.format("projects/{project=*}/zones/{zone_a}%s{zone_b}", invalidDelimiter));
+      thrown.expectMessage(
+          String.format(
+              "parse error: missing or invalid complex resource ID delimiter character in '%s'",
+              String.format("{zone_a}%s{zone_b}", invalidDelimiter)));
+    }
+  }
+
+  @Test
+  public void complexResourceIdMixedSeparators() {
+    // Separate by a mix of delimiters.
+    PathTemplate template =
+        PathTemplate.create("projects/{project}/zones/{zone_a}~{zone_b}.{zone_c}-{zone_d}");
+    Map<String, String> match =
+        template.match(
+            "https://www.googleapis.com/compute/v1/projects/project-123/zones/europe-west3-c~us-east3-a.us-west2-b-europe-west2-b");
+    Truth.assertThat(match).isNotNull();
+    Truth.assertThat(match.get(PathTemplate.HOSTNAME_VAR)).isEqualTo("https://www.googleapis.com");
+    Truth.assertThat(match.get("project")).isEqualTo("project-123");
+    Truth.assertThat(match.get("zone_a")).isEqualTo("europe-west3-c");
+    Truth.assertThat(match.get("zone_b")).isEqualTo("us-east3-a");
+    Truth.assertThat(match.get("zone_c")).isEqualTo("us");
```

Comment:
Look like it uses greedy approach and for the last ambiguous segment (`{zone_c}-{zone_d}`), when provided  with `us-west2-b-europe-west2-b` input it treats the stuff before the first hyphen (`us`) as `zone_c` and everything else as `zone_d`. 

This ambiguity looks like a real problem. Even thought the implementation behaves predictably and reasonably,  looks like `us-west2-b-europe-west2-b` should be split as `zone_c=us-west2-b` and `zone_d=europe-west2-b`.

Should we forbade a character in the body if the same character uses as a splitter?

Reply:
This is a spec question, let's take this discussion over there. @chrisdunelm FYI.

Reply:
Yes, this is a known potential problem.
The spec currently states that it is the responsibility of the API producer to ensure that any split character(s) are not present in the variables. As things are now, I consider this sufficient.
An alternative would be to verify that no parsed variable values contain any of the [used] split character(s); but I think this would be too much,